### ISSUE TITLE
Add multi copy paste (Clipboard history)

### ISSUE THREAD

**Issue by angriness-sleek** [florisboard/florisboard#60]
Like AnySoftKeyboard which makes a list of everything you copied and you can choose to paste anyone

**Comment by D3SOX** [comment]
You mean a clipboard history?

**Comment by angriness-sleek** [comment]
maybe that's what its called...
it's being able ro have more than one copied items in clipboard and being able to paste anyone of those, not just thw latest copied item like most keyboards have

**Comment by patrickgold** [comment]
A clipboard history feature is definitely thinkable of being implemented, as long as it is opt-in. Will add it to the todo list for v0.4.0!

**Comment by X-yl** [comment]
Hey @patrickgold, would you mind if I took a crack at this and #113? It would be a fairly large diff I think, so thought I'd ask before starting.

**Comment by patrickgold** [comment]
@X-yl If you want, you can! It would be best though if you branched off master only after the ` input-logic-rework ` branch is merged in (will do this sometime later this day), because this branch also has a rather large diff which also affects clipboard related commands.

**Comment by X-yl** [comment]
Ah alright, that's fine. Just had a couple questions about the implementation:

- The clipboard history would be accessed through an icon on the clipboard cursor row right? Do you have an icon already or should I make one? 
- How do you want to arrange the clipboard history? In a staggered grid view like gboard or just a regular 
- Just want to make sure, the floris clipboard will never update the android clipboard, but the android clipboard needs to update the floris clipboard (so that copying things from the EditText context menu works). Should the android clipboard then be cleared? If so, there should be a way to force floris to copy to the android clipboard?
- Are there any other notes for the implementation?

**Comment by patrickgold** [comment]
> The clipboard history would be accessed through an icon on the clipboard cursor row right?

This is one way, yes, but it should also be accessible either via a enter popup or the `,` popup for the case that the Smartbar is not visible and optionally also via a swipe action.

> Do you have an icon already or should I make one?

Nope, I don't have a pre-made icon for this. The preferred way is to take one from the Material Icon Library. A hot candidate would be [this](https://material.io/resources/icons/?search=clipboard&icon=assignment&style=outline) icon, which looks similar to the one Gboard uses. If you have a better idea though, I am open to other icons here!

> How do you want to arrange the clipboard history? In a staggered grid view like gboard or just a regular

I think the staggered grid view like Gboard does it looks pretty nice, so I'd go for this one.

> Just want to make sure, the floris clipboard will never update the android clipboard, but the android clipboard needs to update the floris clipboard (so that copying things from the EditText context menu works). Should the android clipboard then be cleared? If so, there should be a way to force floris to copy to the android clipboard, right?

My general thoughts are the following: Have preferences for each sync direction and then allow customization of this sync.

- "Sync Android to Floris Clipboard" (`prefs.clipboard.sync_android_to_floris`): Switch preference, if `true` then the Android clipboard content is synced with the Floris one. Defaults to `true`.
- "Sync Floris to Android Clipboard" (`prefs.clipboard.sync_floris_to_android`): Switch preference, if `true` then the Floris clipboard content is synced with the Android one. Defaults to `false` (because this is a potential privacy risk, but we still give the user the choice to enable syncing if really wanted).

The input events for `CLIPBOARD_CUT`, `CLIPBOARD_COPY` and `CLIPBOARD_PASTE` must of course follow the set behavior. For automatic clipboard clearing I'd have a separate preference for this, which allows to set a time (like 10 sec, 30 sec, 1 min ..),  or disable this completely.

If you have any other clipboard-related preferences, feel free to add them in the `prefs.clipboard` namespace (You need to create this one in the PrefHelper, as it does not exist yet). As for the placement of the clipboard preferences, after the preference UI rewrite in v0.5.0 all clipboard preferences will be on their own page, but for now just put them on the end of the Keyboard preference page.

> Are there any other notes for the implementation?

Just a big thanks that you are willing to implement this rather big feature, really appreciate your contribution! If you have any questions, feel free to ask me any time. :)

**Comment by patrickgold** [comment]
Just so you know I've delayed the merge of the `input-logic-rework` branch to tomorrow because I've found some ultra nasty bugs in the EditorInstance which get annoying when typing fast. I've fixed most of them but there's still a bug open which currently prevents me from merging the branch.

**Comment by X-yl** [comment]
Alright thanks for clarifying. I'll get started soon :+1:

**Comment by X-yl** [comment]
Quick update! I've got the basics working!

![Screenshot_20210310-150358_Evie.jpg](https://user-images.githubusercontent.com/28653555/110619989-ef5d5d80-81b1-11eb-891d-0ecab2ac00f6.jpg)

I'd like some advice though. How would you suggest implementing the context menu that lets you pin an item? (As in like what kind of View/Layout) It appears when you long press a clipboard item. I'm not very experienced with android development so I don't really have many ideas. Here's a screenshot from gboard:

![Screenshot_20210310-151327_GitHub.jpg](https://user-images.githubusercontent.com/28653555/110621168-53345600-81b3-11eb-8fd7-8418af95426f.jpg)

**Comment by patrickgold** [comment]
> Quick update! I've got the basics working!

Great, looks really good!

> I'd like some advice though. How would you suggest implementing the context menu that lets you pin an item? (As in like what kind of View/Layout)

My idea/mock-up for the context menu of an clipboard item:

![image](https://user-images.githubusercontent.com/19412843/110656754-892c0700-81c0-11eb-8c10-4ef33a7235ee.png)
(the font is inaccurate but did not have a better font installed on my machine)

This context menu appears after long pressing an clipboard entry and is always positioned in the center horizontally and above the item vertically, no matter where you long press the clipboard item and can also overlap with the target app's screen if necessary (if the top-most item's context menu is shown, like with the key popups from the first row). This way it is always clear for what item the context menu is and it doesn't cover the whole clipboard manager UI.

**Comment by patrickgold** [comment]
And btw I've merged the `input-logic-rework` branch into master, so master is now up-to-date again and is ready for clean branching off.

**Comment by patrickgold** [comment]
> That's great. Just rebased, there were only a couple conflicts in EditorInstance.

That's great to hear!

> Hope that doesn't accidentally destroy everything lol.

Yes and no xD. This few lines of code are highly fragile and were composed like that after hours of debugging to have a code which executes the copy/cut/paste command for all apps. Your method works for most apps, but some apps just do their own thing and don't wanna behave like they should. Of course you must edit these lines though to catch copy/cut/paste commands for the Floris internal clipboard manager, so it would be best if you mixed your internal update methods with the lines I composed during the rework while respecting the syncing preferences.

**Comment by X-yl** [comment]
Hmm... I'm not sure how that would work? Because your lines currently directly copy/cut directly to the system clipboard, however if the sync options are disabled, the lines should just stay in the FlorisBoard clipboard.

And do you mean the GitJournal thing? because I gave it a quick test and it _seems_ to work. The way I've reimplemented the pasting is using commitText so it bypasses the ctrl+v issue.

You can checkout my (very WIP) code and give it a quick test: https://github.com/X-yl/florisboard/tree/clipboard-stuff

**Comment by patrickgold** [comment]
Ok I've just tried out your clipboard-stuff branch and it looks really good for its state :+1: ! Especially the preferences for the clipboard are very clean and to the point (not like some of my naming ideas for preferences which confuse everyone at the beginning lol).

Regarding the clipboard cut/copy/paste: I just tried it out and it still works in all problematic apps, so I guess your method also works very well, good to know.

**Comment by X-yl** [comment]
> Regarding the clipboard cut/copy/paste: I just tried it out and it still works in all problematic apps, so I guess your method also works very well, good to know.

Okay good, I'm not sure how I would rework it otherwise haha.

Although now I'm facing an issue that I'm not sure how to resolve, help would be appreciated.

I set up a popup to appear using the PopupLayerView, that works fine. The issue is now how do I check that the popup has been clicked? When I click within the keyboard area, there's a touch event sent to the PopupLayerView and the ClipboardHistoryView so I can catch either of those. But when I click outside of the keyboard area, the event is sent directly to the app behind the keyboard instead of to PopupLayerView, and I'm not sure how to change that or work around it.

Thanks for your patience lol, Android development is really not my area of expertise

**Comment by patrickgold** [comment]
> But when I click outside of the keyboard area, the event is sent directly to the app behind the keyboard instead of to PopupLayerView, and I'm not sure how to change that or work around it

That's because of the window insets which are set here: **https://github.com/florisboard/florisboard/blob/e4f5fcf74bb50224b3ad07def501061018221431/app/src/main/java/dev/patrickgold/florisboard/ime/core/FlorisBoard.kt#L569-L582**

Only the touch area which is inside the inset receives touch events, else the event is delegated to the target app. Meaning: you need to recalculate the window insets in this function if a popup is set. I'd suggest that you just set the whole screen as a touch area of FlorisBoard when a clipboard context menu is shown. You can do this by adding this code after line 578:

```
        if (isClipboardContextMenuShown) {
            outInsets?.contentTopInsets = 0
            outInsets?.visibleTopInsets = 0
            return
        }
```

and then you should be able to receive touch input events. Let me know if this works, else I will try to find another solution.

**Comment by X-yl** [comment]
Oh that's perfect, works like a charm. Thanks!

**Comment by X-yl** [comment]
It seems I spoke too soon. There is an issue in some apps that causes most of the UI to go blank when the popup opens. It affects discord, WhatsApp, gmail and probably more. I think we may need a different solution.

https://user-images.githubusercontent.com/28653555/111019947-e6e86b00-83db-11eb-9008-810cf1a72f66.mp4

**Comment by patrickgold** [comment]
You are right I just tested it and it indeed hides the UI. I've found a different solution for onComputeInsets():

```kt
    override fun onComputeInsets(outInsets: Insets?) {
        super.onComputeInsets(outInsets)
        val inputView = this.inputView ?: return
        val inputWindowView = this.inputWindowView ?: return
        // TODO: Check also if the keyboard is currently suppressed by a hardware keyboard
        if (!isInputViewShown) {
            outInsets?.contentTopInsets = inputWindowView.height
            outInsets?.visibleTopInsets = inputWindowView.height
            return
        }
        val visibleTopY = inputWindowView.height - inputView.measuredHeight
        outInsets?.contentTopInsets = visibleTopY
        outInsets?.visibleTopInsets = visibleTopY
        if (isClipboardContextMuneShown) {
            outInsets?.touchableInsets = Insets.TOUCHABLE_INSETS_FRAME
            outInsets?.touchableRegion?.setEmpty()
        }
    }
```

I've tested this new method out in Discord and Gmail and it worked seamlessly, the UI stayed in the background but the touch events went to FlorisBoard instead to the target app.

**Comment by X-yl** [comment]
Awesome! That works. I guess this is pretty much feature complete now? I've probably forgotten something and accidentally caused a few bugs, but I'll make a pull request so you can test it out and review the diff.

**Comment by patrickgold** [comment]
Thanks a lot, really appreciate it! Will try out your changes and review the code changes later today and give feedback on it!

**Comment by patrickgold** [comment]
Ok I finally had the time to try out your changes and must say I am very impressed! The UI is well structured as well as the preferences and the code looks very clean too. Still, I've found some issues I want to share with you (besides the already known ones in your PR):

- When normal-pressing a clipboard item the animation tap-effect is drawn though nothing happens. In terms of UX, this seems like as this press _should_ do something but in reality it doesn't. I suggest that a normal tap pastes the clipboard item like in Gboard.
- For the string resources, I use the naming scheme `<category>__<label>__<sublabel>` (`<sublabel>` is optional) to avoid overlapping namings of strings and to make it more clear for others who read the code which part of the app this string belongs to. I suggest using `clip__<strname>` for your strings (except the pref string, which already follows the convention).
- When committing image content to the target app, you should clear the composing span first in order to avoid possibly crashes or undefined behaviors. You can do this by calling `ic.finishComposingText()` before committing the content.

I really appreciate your work, if you have any questions for your further rework and bug fixing, feel free to ask!

**Comment by patrickgold** [comment]
I've just discovered that there's a third issue which covers the clipboard history/manager stuff: #435. Reading through this request, it seems that your feature-branch covers most if not all of the requested features and you could add it to the list of issues your changes closes.

**Comment by X-yl** [comment]
> - When normal-pressing a clipboard item the animation tap-effect is drawn though nothing happens. In terms of UX, this seems like as this press _should_ do something but in reality it doesn't. I suggest that a normal tap pastes the clipboard item like in Gboard.

Actually the normal tap does actually copy the item to the primary clipboard so it can be pasted. But the fact that you didn't notice does probably mean it's bad UX lol. No worries, I'll change it.

> - For the string resources, I use the naming scheme `<category>__<label>__<sublabel>` (`<sublabel>` is optional) to avoid overlapping namings of strings and to make it more clear for others who read the code which part of the app this string belongs to. I suggest using `clip__<strname>` for your strings (except the pref string, which already follows the convention).
> - When committing image content to the target app, you should clear the composing span first in order to avoid possibly crashes or undefined behaviors. You can do this by calling `ic.finishComposingText()` before committing the content.

Sure, no problem.

> I really appreciate your work, if you have any questions for your further rework and bug fixing, feel free to ask!

Actually I want to ask, so to save the pins, can I use the Room persistence library? Thought I should ask before adding a dependencies.

> I've just discovered that there's a third issue which covers the clipboard history/manager stuff: #435. Reading through this request, it seems that your feature-branch covers most if not all of the requested features and you could add it to the list of issues your changes closes.

Okay, I'll do that when I reopen the PR.

Thanks for being patient lol

**Comment by patrickgold** [comment]
> Actually the normal tap does actually copy the item to the primary clipboard so it can be pasted.

Hmm I absolutely didn't notice that lol. But yeah like I said - directly pasting it to the target app changes something on-screen and gives the user a visual feedback that his press did something.

> Actually I do want to ask something. So to save the pinned items, can I use the Room persistence library? Thought I should ask before adding a dependencies.

Yes you can! The Room persistence library is part of the AndroidX package, which is already listed in the LICENSE screen, so you don't even have to change anything there.

Just a heads up if you use the `Result` type for returning values - I've switched from the `Result<T, E>` class from michealbull to the internal Kotlin Result class. As far as I've seen you don't use it anywhere, but if you intent to do just use the internal Kotlin Result, this avoids that we have build problems after merging.

**Comment by X-yl** [comment]
Okay, thanks. I've started working on it and I've got the basics of the content provider down, but I need to polish it up a bit. I have to say though, the android docs on Content Providers... well they certainly could be better lol.

**Comment by patrickgold** [comment]
> Okay, thanks. I've started working on it and I've got the basics of the content provider down

Great to hear!

> I have to say though, the android docs on Content Providers... well they certainly could be better lol.

The Android documentation is either documented to the last grain of sand or it almost completely lacks on useful documentation I know lol. Especially those godd'mn content URIs with their permissions and SecurityExceptions... Should you need any assistance, feel free to ask :)

**Comment by X-yl** [comment]
Hey now that I have implemented this full content provider thing, it's actually now possible to have an html items on the clipboard. (previously I just converted it to text). So that means I can fairly easily support pasting of html rich text without too much work. Although this does raise the question, how can I get EditorInstance to copy HTML data? Currently the `performClipboardCopy` just uses selection.text which is just plaintext.

**Comment by patrickgold** [comment]
I am fairly well certain that copy/paste of styled text is possible, though I must say the EditorInstance has absolutely not been built for this. If you look at the docs of the `getTextBeforeCursor()` method (https://developer.android.com/reference/android/view/inputmethod/InputConnection#getSurroundingText(int,%20int,%20int)) then you can see that you can pass an additional flag for retrieving a `SpannableString` which has styles attached to it. I am not sure if an Android `SpannableString` is compatible with pasting a HTML formatted text into an app though...

As Gboard seems to also support plain-text and images only, I think copying HTML is either very hard to get right or outright impossible, so for the sake of not over-complicating things we should just support plaintext and not HTML-formatted text.

**Comment by X-yl** [comment]
Hmm actually it seems all the internal `InputConnection` methods take a `CharSequence`, which `Spannable` inherits from. So strictly speaking it _should_ be easy? (Html.fromHtml converts directly from a html string to a Spannable, so that's trivial).

Unfortunately I tried it and it didn't really work very well. There needs to be some debugging and stuff. I tried with the gmail app and underline worked, but bold and italics didn't. So maybe yeah it's best to leave it for a while.

Anyway, I think I'm mostly done now, I'm just going to do some documentation and clean up the code a bit. I think I'll be ready to reopen the PR tomorrow.

**Comment by patrickgold** [comment]
> Anyway, I think I'm mostly done now, I'm just going to do some documentation and clean up the code a bit. I think I'll be ready to reopen the PR tomorrow.

Perfect, your clipboard manager implementation will then be released with v0.3.10-beta01 (I will release v0.3.9 today, which is the last release in the old format of only stable releases, then I will use -betaXX versions as intermediary releases to bring features faster to users who want the latest features)
